homeassistant/components/nissan_leaf.py

```diff
@@ -0,0 +1,338 @@
+            _LOGGER.debug("Firing Refresh on %s "
+                          " as it's charging and the charging interval"
+                          " has passed.",
+                          self.leaf.nickname)
+            result = True
+        elif (self.data[DATA_CLIMATE] is True and
+              self.last_check + climate_interval < now):
+            _LOGGER.debug("Firing Refresh on %s"
+                          " as climate control is on and"
+                          " the interval has passed.",
+                          self.leaf.nickname)
+            result = True
+
+        if result is True:
+            self.refresh_data()
+
+    def refresh_data(self):
+        _LOGGER.debug("Updating Nissan Leaf Data")
+
+        self.last_check = datetime.today()
+
+        battery_response = self.get_battery()
+        _LOGGER.debug("Got battery data for Leaf")
+
+        if battery_response.answer['status'] == 200:
+            if int(battery_response.battery_capacity) > 100:
+                self.data[DATA_BATTERY] = (
+                    battery_response.battery_percent * 0.05
+                )
+            else:
+                self.data[DATA_BATTERY] = battery_response.battery_percent
+
+            self.data[DATA_CHARGING] = battery_response.is_charging
+            self.data[DATA_PLUGGED_IN] = battery_response.is_connected
+            self.data[DATA_RANGE_AC] = battery_response.cruising_range_ac_on_km
+            self.data[DATA_RANGE_AC_OFF] = (
+                battery_response.cruising_range_ac_off_km
+            )
+
+        _LOGGER.debug("Battery Response: ")
+        _LOGGER.debug(battery_response.__dict__)
+
+        climate_response = self.get_climate()
+
+        if climate_response is not None:
+            _LOGGER.debug("Got climate data for Leaf")
+            _LOGGER.debug(climate_response.__dict__)
+            self.data[DATA_CLIMATE] = climate_response.is_hvac_running
+
+        if self.nissan_connect:
+            try:
+                location_response = self.get_location()
+
+                if location_response is None:
+                    _LOGGER.debug("Empty Location Response Received")
+                    self.data[DATA_LOCATION] = None
+                else:
+                    _LOGGER.debug("Got location data for Leaf")
+                    self.data[DATA_LOCATION] = location_response
+
+                    _LOGGER.debug("Location Response: ")
+                    _LOGGER.debug(location_response.__dict__)
+            except:
+                _LOGGER.error("Error fetching location info")
+
+        self.signal_components()
+
+    def signal_components(self):
+        dispatcher_send(self.hass, SIGNAL_UPDATE_LEAF)
+
+    def get_battery(self):
+        request = self.leaf.request_update()
+        battery_status = self.leaf.get_status_from_update(request)
+        while battery_status is None:
+            _LOGGER.debug("Battery data not in yet.")
+            time.sleep(5)

```

Comment:
Don't use sleep inside sync contexts as it blocks threads. It's ok to sleep within an async context, however, it seems weird to keep polling indefinitely? It should give up after a few times to make we don't break the 12v battery…

```python
async def async_get_battery(self):
    request = await self.hass.async_add_job(self.leaf.request_update)
    for i in range(5):
        if i > 0:
            await asyncio.sleep(5)

        battery_status = await self.hass.async_add_job(self.leaf.get_status_from_update(request))
        if battery_status is not None:
            break

    return battery_status
```

Reply:
I had been using async but was told that will block the main thread instead. Nissan's servers are really bad so they can be pretty sluggish sometimes.

The reason we sleep here is because the initial request will ask for the battery from the car, the requests in the while loop are then asking Nissan's servers if the car has updated them with the latest battery data. It's a pretty poor implementation by them - Tesla constantly phone home so you don't have these issues but Nissan's systems were coded by monkeys.

Reply:
It's ok to use async as long as you offload I/O to the executor (a thread worker pool). That is done in my example by passing the methods that do I/O (like `request_update`) to `hass.async_add_job` and awaiting that. Once the function is executed, the async context will resume.